File: src/engine/Component/Rigidbody.jl
module RigidbodyModule
    using ..Component.JulGame
    using ..Component.JulGame.Math: _Vector2
    import ..Component
    export Rigidbody
    struct Rigidbody
        mass::Float64
        useGravity::Bool

        function Rigidbody(;mass::Float64 = 1.0, useGravity::Bool = true)
            return new(mass, useGravity)
        end
    end

    export InternalRigidbody
    mutable struct InternalRigidbody 
        acceleration::_Vector2{Float64}
        drag::Float64
        grounded::Bool
        mass::Float64
        offset::_Vector2{Float64}
        parent::Any
        useGravity::Bool
        velocity::_Vector2{Float64}

        function InternalRigidbody(parent::Any; mass::Float64 = 1.0, useGravity::Bool = true)
            return new(
                _Vector2{Float64}(),
                0.1,
                false,
                mass,
                _Vector2{Float64}(),
                parent,
                useGravity,
                _Vector2{Float64}(0.0, 0.0),
            )
        end
    end

    function Component.update(this::InternalRigidbody, dt)
        dt = clamp(dt, 0, .5)
        velocityMultiplier = _Vector2{Float64}(1.0, 1.0)
        transform = this.parent.transform
        currentPosition = transform.position
        
        newPosition = transform.position + this.velocity*dt + this.acceleration*(dt*dt*0.5)
        if this.grounded
            newPosition = _Vector2{Float64}(newPosition.x, currentPosition.y)
            velocityMultiplier = _Vector2{Float64}(1.0, 0.0)
        end
        newAcceleration = Component.apply_forces(this)
        newVelocity = this.velocity + (this.acceleration+newAcceleration)*(dt*0.5)

        Component.set_position(transform, newPosition)
        set_velocity(this, newVelocity * velocityMultiplier)
        this.acceleration = newAcceleration

        if this.parent.collider != C_NULL
            Component.check_collisions(this.parent.collider)
        end
    end

    function Component.apply_forces(this::InternalRigidbody)
        gravityAcceleration = _Vector2{Float64}(0.0, this.useGravity ? GRAVITY : 0.0)
        dragForce = 0.5 * this.drag * (this.velocity * this.velocity)
        dragAcceleration = dragForce / this.mass
        return gravityAcceleration - dragAcceleration
    end

    function Component.get_velocity(this::InternalRigidbody)
        return this.velocity
    end

    """
    add_velocity(this::Rigidbody, velocity::Math.Vector2f)

    Add the given velocity to the Rigidbody's current velocity. If the y-component of the velocity is positive, set the `grounded` flag to false.
    
    # Arguments
    - `this::Rigidbody`: The Rigidbody component to set the velocity for.
    - `velocity::Math.Vector2f`: The velocity to set.
    """
    function add_velocity(this::InternalRigidbody, velocity::_Vector2{Float64})
        this.velocity = this.velocity + velocity
        if(velocity.y < 0)
            this.grounded = false
            if this.parent.collider != C_NULL
                this.parent.collider.currentRests = []
            end
        end
    end
    export add_velocity
    
    """
    set_velocity(this::Rigidbody, velocity::Math.Vector2f)

    Set the velocity of the Rigidbody component.

    # Arguments
    - `this::Rigidbody`: The Rigidbody component to set the velocity for.
    - `velocity::Vector2f`: The velocity to set.
    """
    function set_velocity(this::InternalRigidbody, velocity::_Vector2{Float64})
        this.velocity = velocity
        if(velocity.y < 0)
            #this.grounded = false
        end
    end
    export set_velocity

    function Component.duplicate(this::InternalRigidbody, parent::Any)
        newRigidbody = InternalRigidbody(parent, mass=this.mass, useGravity=this.useGravity)
        newRigidbody.acceleration = this.acceleration
        newRigidbody.drag = this.drag
        newRigidbody.grounded = this.grounded
        newRigidbody.mass = this.mass
        newRigidbody.offset = this.offset
        return newRigidbody
    end
end

File: ts/_generated/src/engine/Component/Rigidbody.ts
export {}
import { clamp } from "../../../../src/engine/core/juliaHelpers";
import { vecAdd, vecSub, vecMul, vecDiv, vecNeg } from "../../../../src/engine/core/vectorOps";


    // using ..Component.JulGame
    // import ..Component
    
    class InternalRigidbody { 
        acceleration: Vector2f
        drag: number
        grounded: boolean
        mass: number
        offset: Vector2f
        parent: any
        useGravity: boolean
        velocity: Vector2f

        constructor(parent: any, mass: number = 1.0, useGravity: boolean = true) {
            
            
            this.acceleration = {x: 0, y: 0}
            this.drag = 0.1
            this.grounded = false
            this.mass = mass
            this.offset = {x: 0, y: 0}
            this.parent = parent
            this.useGravity = useGravity
            this.velocity = {x: 0.0, y: 0.0}

        }
    }

    function Component_update(self: InternalRigidbody, dt: number) {
        dt = clamp(dt, 0, .5)
        let velocityMultiplier = {x: 1.0, y: 1.0}
        let transform = self.parent.transform
        let currentPosition = transform.position
        
        let newPosition = vecAdd(vecAdd(transform.position, vecMul(self.velocity, dt)), vecMul(self.acceleration, vecMul(vecMul(dt, dt), 0.5))) as Vector3f
        if (self.grounded) {
            newPosition = {x: newPosition.x, y: currentPosition.y, z: newPosition.z}
            velocityMultiplier = {x: 1.0, y: 0.0}
        }
        let newAcceleration = Component_apply_forces(self)
        let newVelocity = vecAdd(self.velocity, vecMul(vecAdd(self.acceleration, newAcceleration), vecMul(dt, 0.5))) as Vector2f

        transform.position = newPosition
        self.velocity = vecMul(newVelocity, velocityMultiplier) as Vector2f
        self.acceleration = newAcceleration

        if (self.parent.collider != null) {
            Component_check_collisions(self.parent.collider)
        }
    }

    function Component_apply_forces(self: InternalRigidbody) {
        let gravityAcceleration = {x: 0.0, y: self.useGravity ? (globalThis as any).JulGame.GRAVITY : 0.0}
        let dragForce = vecMul(vecMul(0.5, self.drag), vecMul(self.velocity, self.velocity)) as Vector2f
        let dragAcceleration = vecDiv(dragForce, self.mass) as Vector2f
        return vecSub(gravityAcceleration, dragAcceleration) as Vector2f
    }

    function Component_get_velocity(self: InternalRigidbody) {
        return self.velocity
    }

    /*
    add_velocity(this, velocity)

    Add the given velocity to the Rigidbody's current velocity. If the y-component of the velocity is positive, set the `grounded` flag to false.
    
    // Arguments
    - `this`: The Rigidbody component to set the velocity for.
    - `velocity`: The velocity to set.
    */
    function add_velocity(self: InternalRigidbody, velocity: Vector2f) {
        self.velocity = vecAdd(self.velocity, velocity) as Vector2f
        if (velocity.y < 0) {
            self.grounded = false
            if (self.parent.collider != null) {
                self.parent.collider.currentRests = []
            }
        }
    }
    

    function Component_duplicate(self: InternalRigidbody, parent: any) {
        let newRigidbody = new InternalRigidbody(parent, self.mass, self.useGravity)
        newRigidbody.acceleration = self.acceleration
        newRigidbody.drag = self.drag
        newRigidbody.grounded = self.grounded
        newRigidbody.mass = self.mass
        newRigidbody.offset = self.offset
        return newRigidbody
    }

export { InternalRigidbody }
